File: Sources/Mobi2EpubTransferCore/DownloadedBookValidator.swift
import Foundation

public enum DownloadedBookValidationError: Error, Equatable, LocalizedError {
    case unsupportedOrDisguisedFile

    public var errorDescription: String? {
        "The download is not a supported EPUB, MOBI, AZW, comic archive, or PDF file."
    }
}

public enum DownloadedBookValidator {
    public static func validatedExtension(at url: URL, suggestedExtension: String) throws -> String {
        let handle = try FileHandle(forReadingFrom: url)
        defer { try? handle.close() }
        let header = try handle.read(upToCount: 4096) ?? Data()
        let ext = suggestedExtension.lowercased()
        let zip = header.starts(with: [0x50, 0x4B])
        let pdf = String(data: header.prefix(5), encoding: .ascii) == "%PDF-"
        let rar = String(data: header.prefix(4), encoding: .ascii) == "Rar!"
        let mobi = header.count >= 68 && String(data: header[60..<68], encoding: .ascii)?.contains("BOOKMOBI") == true
        let epub = zip && header.range(of: Data("application/epub+zip".utf8)) != nil
        let prefix = String(data: header.prefix(256), encoding: .utf8)?.lowercased() ?? ""
        let looksLikeHTML = prefix.contains("<!doctype html") || prefix.contains("<html")
        let executableMagic: [[UInt8]] = [[0xCF, 0xFA, 0xED, 0xFE], [0xCE, 0xFA, 0xED, 0xFE], [0x4D, 0x5A]]
        guard !looksLikeHTML, !executableMagic.contains(where: { header.starts(with: $0) }) else {
            throw DownloadedBookValidationError.unsupportedOrDisguisedFile
        }
        switch ext {
        case "epub" where epub: return "epub"
        case "cbz" where zip: return "cbz"
        case "zip" where zip: return "zip"
        case "pdf" where pdf: return "pdf"
        case "cbr" where rar: return "cbr"
        case "mobi" where mobi: return "mobi"
        case "azw" where mobi: return "azw"
        case "azw3" where mobi: return "azw3"
        default:
            if epub { return "epub" }
            if pdf { return "pdf" }
            if rar { return "cbr" }
            if mobi { return ext.isEmpty ? "mobi" : ext }
            if zip { return "zip" }
            throw DownloadedBookValidationError.unsupportedOrDisguisedFile
        }
    }
}

File: Windows/src/MobiVerse.Core/DownloadedBookValidator.cs
using System.Text;

namespace MobiVerse.Core;

public sealed class DownloadedBookValidationException() : Exception(
    "The download is not a supported EPUB, MOBI, AZW, comic archive, or PDF file.");

public static class DownloadedBookValidator
{
    public static string ValidatedExtension(string path, string suggestedExtension)
    {
        using var stream = File.OpenRead(path);
        var header = new byte[Math.Min(4096, (int)Math.Min(stream.Length, 4096))];
        _ = stream.Read(header, 0, header.Length);
        var extension = suggestedExtension.TrimStart('.').ToLowerInvariant();
        var zip = StartsWith(header, [0x50, 0x4b]);
        var pdf = Encoding.ASCII.GetString(header.Take(5).ToArray()) == "%PDF-";
        var rar = Encoding.ASCII.GetString(header.Take(4).ToArray()) == "Rar!";
        var mobi = header.Length >= 68 && Encoding.ASCII.GetString(header, 60, 8).Contains("BOOKMOBI", StringComparison.Ordinal);
        var epub = zip && Encoding.ASCII.GetString(header).Contains("application/epub+zip", StringComparison.Ordinal);
        var prefix = Encoding.UTF8.GetString(header.Take(256).ToArray()).ToLowerInvariant();
        var looksLikeHtml = prefix.Contains("<!doctype html") || prefix.Contains("<html");
        var executable = StartsWith(header, [0xcf, 0xfa, 0xed, 0xfe]) ||
                         StartsWith(header, [0xce, 0xfa, 0xed, 0xfe]) || StartsWith(header, [0x4d, 0x5a]);
        if (looksLikeHtml || executable) throw new DownloadedBookValidationException();

        return extension switch
        {
            "epub" when epub => "epub",
            "cbz" when zip => "cbz",
            "zip" when zip => "zip",
            "pdf" when pdf => "pdf",
            "cbr" when rar => "cbr",
            "mobi" when mobi => "mobi",
            "azw" when mobi => "azw",
            "azw3" when mobi => "azw3",
            _ when epub => "epub",
            _ when pdf => "pdf",
            _ when rar => "cbr",
            _ when mobi => string.IsNullOrWhiteSpace(extension) ? "mobi" : extension,
            _ when zip => "zip",
            _ => throw new DownloadedBookValidationException()
        };
    }

    private static bool StartsWith(byte[] value, byte[] prefix) =>
        value.Length >= prefix.Length && value.AsSpan(0, prefix.Length).SequenceEqual(prefix);
}
